### ISSUE TITLE
inclusion of WinFlex changes (drop need for fork/wait and posix conversion function htonl/htons)

### ISSUE THREAD

**Issue by GitMensch** [westes/flex#367]
https://github.com/lexxmark/winflexbison/tree/master/flex/src includes some changes for non-GCC (obviously especially WIN32) compilers.

What do you think about integrating them (after cleanup)?

A good starting example seems to be filter.c (and flexdef.h) to use a different signature of `struct filter` and `filter_apply_chain ()`. These changes could be integrated if one or both functions `fork()`/`wait()` are not available (at least with 2.6.4 there is a check in configure for them but neither a configure time failure nor a `#ifdef HAVE_FORK` in the code).

Similar could be done for tables.c and `htonl` / `htons` (which seem to have no check during configure).

So... was a discussion about this before?

**Comment by GitMensch** [comment]
Note: if this is agreed upon I'd add some PR with getting similar changes in that are needed to compile and run with OrangeC and VS (as it seems to be the main target of the WinFlexBison changes).

**Comment by Explorer09** [comment]
@GitMensch I have discovered the non-portability of htonl / htons, but I'm not sure how WinFlexBison solves it. (htonl and htons are used in skeleton, so adding a check of them in configure is useless. We need other workaround for the case.) Can you give a better pointer?

Regarding the others, I can't comment at the moment.

**Comment by Explorer09** [comment]
@GitMensch 
Maybe I was wrong about htons/htonl pair, but their inverse, `ntohs` and `ntohl`, have to be included in the skeleton as it is part of the decoder of the flex table file format.

I believe making htons/htonl macro defines would cause more portability problem than it solves, since (I think) some compilers would recognize htons/htonl functions and optimize them into inline assembly without generating function calls. Repeating what a compiler would do in a header is a bad idea.

I think with portability to Windows, including <winsock.h> would be the way to go. But for even more exotic platforms, I don't know if we would need this naive fallback:

```c
short htons(short val) {
    unsigned char arr[sizeof(short)] = {0};
    int i;
    for (i = sizeof(short) - 1; i >= 0; i--) {
        arr[i] = (unsigned char)(val);
        val >>= CHAR_BIT;
    }
    return (*(short *)arr);
}
```

(Yes, you get it. This is the most portable htons implementation within ANSI C constraints.)

**Comment by westes** [comment]
If we're going to provide an implementation of a function, we should use the libobj mechanism.

On Sunday, 24 June 2018,  3:39 am -0700, "Kang-Che Sung (宋岡哲)" <notifications@github.com> wrote:

> @GitMensch 
> Maybe I was wrong about htons/htonl pair, but their inverse, `ntohs` and `ntohl`, have to be included in the skeleton as it is part of the decoder of the flex table file format.
> 
> I believe making htons/htonl macro defines would cause more portability problem than it solves, since (I think) some compilers would recognize htons/htonl functions and optimize them into inline assembly without generating function calls. Repeating what a compiler would do in a header is a bad idea.
> 
> I think with portability to Windows, including <winsock.h> would be the way to go. But for even more exotic platforms, I don't know if we would need this naive fallback:
> 
> ```c
> short htons(short val) {
>     unsigned char arr[sizeof(short)] = {0};
>     int i;
>     for (i = sizeof(short) - 1; i >= 0; i--) {
>         arr[i] = (unsigned char)(val);
>         val >>= CHAR_BIT;
>     }
>     return (*(short *)arr);
> }
> ```
> 
> (Yes, you get it. This is the most portable htons implementation within ANSI C constraints.)
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/westes/flex/issues/367#issuecomment-399746765

-- 
Will Estes
westes575@gmail.com

**Comment by Explorer09** [comment]
@westes I think the licensing problem (it's LGPL) already prevents us from including libcob's header directly info flex's source.

**Comment by westes** [comment]
libcob?

On Sunday, 24 June 2018,  6:29 am -0700, "Kang-Che Sung (宋岡哲)" <notifications@github.com> wrote:

> @westes I think the licensing problem (it's LGPL) already prevents us from including libcob's header directly info flex's source.
> 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/westes/flex/issues/367#issuecomment-399756660

-- 
Will Estes
westes575@gmail.com

**Comment by Explorer09** [comment]
@westes I was talking about libcob that @GitMensch mentioned. And I had no idea what is "libobj mechanism" you are talking about.

**Comment by westes** [comment]
libobj as handled by automake. It's the way to provide portability functions when absent from the system at compile time.

On Sunday, 24 June 2018,  6:36 am -0700, "Kang-Che Sung (宋岡哲)" <notifications@github.com> wrote:

> @westes I was talking about libcob that @GitMensch mentioned. And I had no idea what is "libobj mechanism" you are talking about.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/westes/flex/issues/367#issuecomment-399757294

-- 
Will Estes
westes575@gmail.com

**Comment by Explorer09** [comment]
@westes 
Gosh. There're many similarly named "libobj" library out there on the web, so it was not easy to see which one you were talking about.
Well, I could make a PR about adding `htons` and `htonl` to automake $LIBOBJS, but that wouldn't solve the problem of `ntohs` and `ntohl` in the generated scanner code.
I'm not sure what to do with the latter case. Adding a fallback version of `yyflex_ntohs` and `yyflex_ntohl` into the skeleton?

**Comment by lexxmark** [comment]
Hello everyone,

when I ported flex to Windows I used htons/htonl functions from <winsock.h>.
then I decided to remove dependency from Ws2_32.dll by replacing htons/htonl with macro (because wintel has only one byte order).

I can easily revert back my "improvement" and use htons/htonl again.

ntohs/ntohl are used in generated code only. I didn't changed this code so parsers are dependent from Ws2_32.dll under windows now. And I think it's not a big deal.

In summary, I don't think ntohs/ntohl is a problem at all.

**Comment by GitMensch** [comment]
@lexxmark Please revert this change to use functions from winsock.h. I'll then try to create a PR to this repo which includes the checks for `#if defined (HAVE_SYS_WAIT_H) && defined (HAVE_FORK)` and `HAVE_WINSOCK_H` (including configure.ac to set these).

This should be a good start to minimize the differences between this repo and your port.

**Comment by Explorer09** [comment]
@lexxmark
If it's me, I would like to have htons/htonl function names preserved. The reason is they're POSIX, even though they're not part of standard (ISO) C (I wish it become so, or at least reserve the names, so that when people refer to `htonl` they will always mean this function).
I've read your code ([this example](https://github.com/lexxmark/winflexbison/blob/v2.5.15/flex/src/tables.c)), and I don't like the idea of hard-coding `SWAP_BYTE_ORDER_UINT()`, as it would bring _more_ portability problems than it solves.

If you read my post above this issue report, I have presented a code that does endian-independant appraroch to byte swapping (or [see this article on the web](https://commandcenter.blogspot.com/2012/04/byte-order-fallacy.html)). If we are to add a htons or htonl in Flex, it would be that most portable version. Otherwise, I would better tell compilers to fix their optimization rather than implement a workaround that would always be inferior to a machine instruction.

**Comment by GitMensch** [comment]
@Explorer09 lexxmark wrote about using htons/htonl in winflexbison again (with additional winsock.h and link to Ws2_32.dll). As soon as this is done I'll do something similar for flex using defines and checks in configure.ac + PR here.

**Comment by lexxmark** [comment]
I've reverted htons/htonl (see my [commit](https://github.com/lexxmark/winflexbison/commit/c42b8261d86789c431fd844263db33d6c5eb057f))
